internal/host/plugin/job_set_sync.go

```diff
@@ -135,6 +135,13 @@ func (r *SetSyncJob) Description() string {
 }
 
 func nextSync(j scheduler.Job) (time.Duration, error) {
+	// FIXME: This function should query the database and figure out the right
+	// time, however, at the time of writing we don't have an update function
+	// for sets which means we cannot call UpdateJobNextRunInAtLeast if a value
+	// changes. So rather than use the DB query, we simply check every five
+	// seconds. The actual job will ensure that only sets needing updates will
+	// be updated.
+	return 15 * time.Second, nil // Scheduler only wakes up every minute anyways
```

Comment:
The comment says five seconds, the code says 15 seconds.

Another thing to keep in mind is that while the scheduler wakes up only every minute that is per controller.  If there are multiple controllers running they could be running in a way to have at least 1 of the controller's schedulers waking up more frequently than that.  The running of the job can short circuit itself by running just a single query but that query relies on a column that doesn't have an explicit index so it could potentially hit every plugin host set in the system every time.  Is 15 seconds to frequent of a time for that?  

Thinking about this more also, the value cannot change from what it is set to at create time because there is no update set call yet so there is no risk of miscalculating the next sync job call.

Also, although right now it looks like it's only called at the end of a successful run of a job to decide when it should be run again, if this function is ever called in the flow to determine if the job should run this could starve the job since it will always be 15 seconds away from being ready to run (but hopefully the fixme will be fixed by then)

Reply:
> The comment says five seconds, the code says 15 seconds.

Forgot to update the comment, apologies.

> The running of the job can short circuit itself by running just a single query but that query relies on a column that doesn't have an explicit index so it could potentially hit every plugin host set in the system every time. Is 15 seconds to frequent of a time for that?

The query doesn't seem like it should be very expensive, so running once a minute seems like it should be fine -- even once a minute per controller. Do you think we should be adding an index?

> Thinking about this more also, the value cannot change from what it is set to at create time because there is no update set call yet so there is no risk of miscalculating the next sync job call.
True, but we'll have that by release time...but that's actually still a problem. Imagine you have three controllers and sync is disabled. You run update on one, which informs the scheduler that it should run sooner -- but the other two controllers don't know this. So now only one controller will ever be able to run that job because the scheduler never checks for a new value until <never> happens. (This may well already affect the Vault jobs -- @louisruch ? ) If that controller goes down, the job may never be run(?) The way out of that is to have the scheduler check periodically anyways.

Generally speaking I'm not sure at this point if it makes sense to calculate next run times rather than have jobs define an interval for checking and just run the queries on that interval to see if anything needs to be run now. I'd rather have more queries than starvation.

Reply:
Everything updated is persisted to the db, if you call `UpdateJobNextRunInAtLeast` the scheduler updates the next run for that job in the database so all controllers would see the change and which ever controller wakes up when/or after the job is scheduled to run would run that job.

Reply:
@louisruch Ahh OK, thanks.